src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceDataRegistryKey.cs

```diff
@@ -0,0 +1,142 @@
+
+        private PerformanceDataRegistryKey(SafeRegistryHandle hkey)
+        {
+            _hkey = hkey;
+        }
+
+        public static PerformanceDataRegistryKey OpenRemoteBaseKey(string machineName)
+        {
+            // connect to the specified remote registry
+            SafeRegistryHandle foreignHKey = null;
+            int ret = Interop.Advapi32.RegConnectRegistry(machineName, new SafeRegistryHandle(new IntPtr(PerformanceData), ownsHandle: false), out foreignHKey);
+
+            if (ret == Interop.Errors.ERROR_DLL_INIT_FAILED)
+            {
+                // return value indicates an error occurred
+                throw new ArgumentException(SR.Format(SR.Arg_DllInitFailure, machineName));
+            }
+
+            if (ret != 0)
+            {
+                Win32Error(ret, null);
+            }
+
+            if (foreignHKey.IsInvalid)
+            {
+                // return value indicates an error occurred
+                throw new ArgumentException(SR.Format(SR.Arg_RegKeyNoRemoteConnect, machineName));
+            }
+
+            return new PerformanceDataRegistryKey(foreignHKey);
+        }
+
+        public static PerformanceDataRegistryKey OpenLocal()
+        {
+            var key = new SafeRegistryHandle(new IntPtr(PerformanceData), ownsHandle: true);
+            return new PerformanceDataRegistryKey(key);
+        }
+
+        public byte[] GetValue(string name, bool usePool)
+        {
+            int size = 65000;
+            int sizeInput = size;
+
+            int ret;
+            int type = 0;
+            byte[] data = CreateBlob(size, usePool);
+            while (Interop.Errors.ERROR_MORE_DATA == (ret = Interop.Advapi32.RegQueryValueEx(_hkey, name, lpReserved: null, ref type, data, ref sizeInput)))
+            {
+                if (size == int.MaxValue)
+                {
+                    ReleaseData(data, usePool);
+
+                    // ERROR_MORE_DATA was returned however we cannot increase the buffer size beyond Int32.MaxValue
+                    Win32Error(ret, name);
+                }
+                else if (size > (int.MaxValue / 2))
+                {
+                    // at this point in the loop "size * 2" would cause an overflow
+                    size = int.MaxValue;
+                }
+                else
+                {
+                    size *= 2;
+                }
+                sizeInput = size;
+
+                ReleaseData(data, usePool);
+                data = CreateBlob(size, usePool);
+            }
+
+            if (ret != 0)
+            {
+                ReleaseData(data, usePool);
+                Win32Error(ret, name);
+            }
+
+            return data;
+        }
+
+        public void ReleaseData(byte[] data, bool usePool = true)
+        {
+            if (usePool)
+            {
+#if DEBUG
+                // Fill array by garbage to detect early situations when this array is used after return
```

Comment:
I do not think we should be adding ad-hoc fills like this under ifdef. If we want to make this diagnosable, it should be

Reply:
@jkotas did you comment get cut off?

Reply:
Do you prefer to have this under `#if DEBUG` in the ConfigurableArrayPool itself -- the danger of course is release-only bugs, but it seems hard to imagine in this case? Or behind `#if DEBUG` as well as some environment variable a dev could set?

Reply:
Ignore this. I started writing this, but then I have decided that it is probably worth it here given the complexity of the code. I thought I have hit cancel, but apparently not.

Reply:
> Do you prefer to have this under #if DEBUG

The array pool diagnostic - once we feel enough pain to add it - should in the release builds, controlled by environment variable. No different from how malloc diagnostic works https://www.bing.com/search?q=MALLOC_CHECK_ .

Reply:
@danmosemsft, do you mean such kind of check? or i've done something wrong?

Reply:
@MaxShoshin please remove this `#if DEBUG` block. My bad. @jkotas suggests we do this centrally. That would be a separate PR by someone, tracked in dotnet/coreclr#9885.